### PR TITLE
add array::from_ref

### PR DESCRIPTION
**Pull request by lcnr**
mirrors the methods in `std::slice` with the same name.

I guess this method previously didn't exist as there was close to no reason to create an array of size `1`.
This will change due to const generics in the near future.

### CODE REVIEW

File: library/core/src/array/mod.rs [hunk 1]
```diff
@@ -19,6 +19,20 @@ mod iter;
 #[unstable(feature = "array_value_iter", issue = "65798")]
 pub use iter::IntoIter;
 
+/// Converts a reference to `T` into a reference to an array of length 1 (without copying).
+#[unstable(feature = "array_from_ref", issue = "none")]
```
Comment (SimonSapin) [inline library/core/src/array/mod.rs:23]: Please file a tracking issue and replace `none` here and on `from_mut` below before merging.

Comment (lcnr) [reply library/core/src/array/mod.rs:23]: yeah, didn't want to open a tracking issue if we aren't yet sure about adding merging that PR.

Considering that you expressed some interest in adding these methods, I will just go ahead and do so now

**Comment by SimonSapin** [comment]
This is new unstable safe API that would be very easy to re-implement, but require `unsafe`. It is rather niche, but follows the precedent of similar functions in `std::slice`. At least one person asked about exactly this: https://www.reddit.com/r/rust/comments/ixqzkf/are_there_any_gotchas_to_converting_t_into_t_1/

@rust-lang/libs Does anyone feel this should *not* be merged and eventually stabilized?

**Comment by sfackler** [comment]
I'm not really opposed, but why will const generics change the usefulness of this method?

**Comment by lcnr** [comment]
const generics makes working with arrays easier, which will cause more methods to both create and use arrays.

**Comment by emilazy** [comment]
Just wanted to point out that `slice::from_{ref,mut}` can both be reimplemented more simply in terms of `array::from_{ref,mut}` due to `&[T; 1]` automatically decaying to `&[T]`, so this API can be added with no net increase of `unsafe` usage and arguably decreases the subtlety of the safety justifications.

**Comment by SimonSapin** [comment]
Good point @emilazy.

@lcnr, could you make the change in this PR? r=me with that.

**Pull request merged**